doc/runtime-information.rst

```diff
@@ -48,47 +48,46 @@ case)::
     from os import path
     path_to_dat = path.abspath(path.join(path.dirname(__file__), 'file.dat'))
 
-In the bundled main script itself the above might not work, as it is unclear
-where it resides in the package hierarchy. So in when trying to find data files
-relative to the main script, ``sys._MEIPASS`` can be used. The following will
-get the path to a file ``other-file.dat`` next to the main script if not
-bundled and in the bundle folder if it is bundled::
+In the main script (the ``__main__`` module) itself, the ``__file__``
+variable contains path to the script file. In Python 3.8 and earlier,
+this path is either absolute or relative (depending on how the script
+was passed to the ``python`` interpreter), while in Python 3.9 and later,
+it is always an absolute path. In the bundled script, the |PyInstaller|
+|bootloader| always sets the ``__file__`` variable inside the ``__main__``
+module to the absolute path inside the bundle directory, as if the
+byte-compiled entry-point script existed there.
+
+For example, if your entry-point script is called ``program.py``, then
+the ``__file__`` attribute inside the bundled script will point to
+``sys._MEIPASS + 'program.pyc'``. Therefore, locating a data file relative
+to the main script can be either done directly using ``sys._MEIPASS`` or
+via the parent path of the ``__file__`` inside the main script. In earlier
+versions of |PyInstaller|, this was possible only by using ``sys._MEIPASS``,
+but now the preferred way should by using ``__file__``, as it will work
+in both non-bundled and bundled case.
+
```

Comment:
We probably should slip in a
```rst
.. versionchanged:: 5.0

    Formally, the ``__file__`` attribute of the top level ``__main__``
    script was set to only its basename rather than its full path.

```

in there somewhere (assuming 5.0 really is our next release).

Reply:
Putting the historic details into a separate versionchanged block does indeed look better.